Task #3: Add range parsing for condition evaulation

daahl opened the pull request:
First regarding the many commits, it's because I merged my dev branch to this branch.

This pull request includes the implementation of a parsing function, that takes the range condition and parses it into a string which can later be evaluated using eval().

Note that while all general cases are implemented, there are still a handful of special cases left to implement. I'm opening the pull request now to give you more time to leave feedback.

hobler commented:
There are a few conditions which (probably) will not work:

- Text "No condition parsed" cannot be evaluated.
- Boolean (logical) variables are expected to be entered as T or F, but the conditions check on 'true' or 'false'
- "any real number" can be translated into "type(...) in [int, float]"
- Some checks on equality are written with "=" instead of "=="
- Checking two vectors on being normal to each other should not be done by checking the dot product on being equal to 0. This will not work because of numerical errors. Instead, request abs(v.w)<0.001*max(v.v,w.w) where "." denotes the dot product.

Apart from this, there is a problem with some of the special cases. Since they are special, it cannot be guaranteed that they will correctly evaluate if the manual is changed. But they should never evaluate to False if they should evaluate to True according to the new manual. As a compromise, they may evaluate to True if they should evaluate to False. It would be good, if a warning could be printed if that happens.

daahl commented:
I made the requested changes:

- Parameters without condition now return None type instead of "No condition parsed". This can later be checked before running eval()
- Boolean parameters are now assumed to be capital T or F, and are evaluated with "... in ['T', 'F']"
- "Any real number" is evaluated using the suggested fix.
- Fix error with "=" being used instead of "==".
- Dot product is now evaluated using the suggested fix.
- Added warning message and default True evaluation on parameters in special cases with hard-coded values, for parameters where I interpreted it as being useful.

hobler commented:
- I think the special-case strings you return from `get_range_condition()` have syntax errors due to the use of the single quotes (but I don't know how you are going to use them).
- The text "... value accepted, but allowed range might have changed." might be clearer if replaced by "... value accepted, but might be out of range."
- In `&geometry` `XPER` the range `XPER(2)>XPER(1)` (Fortran indexing/natural language) should translate to `XPER[1]>XPER[0]` (Python indexing). The same applies to `&ions` `XINIT`, `YINIT`, `ZINIT`.
- The condition abs(v.w)<0.001\*max(v.v,w.w) should be replaced by abs(v.w)<0.001\*(v.v\*w.w)\*\*0.5 (sorry, my mistake in my last comment)
- The list of chemical elements is incomplete. The first 92 elements should be predefined.

What is your plan for completing Task 4?

daahl commented:
- Special case syntax errors:
    - Fixed multiple instances of missing and/or misplaced single quotation marks.
    - Idea is that the string in “conditions” can be used in an eval() function. The parameters name, e.g. “DENSITY” would first have to be replaced by the parameters actual value, before the string can be evaluated using eval(). It could look like e.g. eval(condition, {“DENSITY”:  DENSITY})
- Clearer warning text:
    - Changed the message as requested.
- Use Python indexing:
    - Changed indexing also for other mentions of XINIT and YINIT.
- Correct use of dot product:
    - Corrected as requested.
- The list of chemical elements is incomplete. The first 92 elements should be predefined.
    - Added the missing elements.
    
    
Regarding Task 4 please see my email.
